Rerun failed tests automatically and detect flaky tests

**Description**

- Use settings in surefire to rerun failing tests. http://maven.apache.org/surefire/maven-surefire-plugin/examples/rerun-failing-tests.html
- Scan the log for flaky tests
- Fail build when flaky tests were found
- collect names of flaky tests and display them prominently

@menski  I could use your help here. I googled my way to a PoC, but I struggle to get it running on the Zeebe Pipeline.

The PoC is here: https://github.com/pihme/jenkinsTestBed/tree/rerun-failed-tests

It works as follows:
- `runTest.sh` runs the tests and then scans the log for flaky tests; if it finds them it collects the interesting lines, extracts the test names and writes them into a file, and fails the stage of the build
- `Jenkinsfile` then looks for the file, and if it's there writes the content into the build description

In Zeebe the second step seems to be not working. Maybe it's because of the "container" tag and the file I write isn't visible to the world outside.

Besides that there is potential for improvement:
- in the PoC I didn't capture the exit code of the original maven run. This would be important to mark the stage as failed if it's an ordinary test failure
- I do the extraction of the tests in two steps: first I use grep to look for the lines I am interested in, and then I use awk to extract a certain "column". I guess both steps could be combined into an elegant regex with capture groups
- now that I think of it, the code will also report tests that always fail as flaky. It needs a better matcher:
```
[WARNING] Flakes: 
[WARNING] com.github.pihme.jenkinstestbed.module1.FlakyTest.shouldFailOnceInTwoInvocations
[ERROR]   Run 1: FlakyTest.shouldFailOnceInTwoInvocations:22 failed
[ERROR]   Run 2: FlakyTest.shouldFailOnceInTwoInvocations:22 failed
[INFO]   Run 3: PASS
```

vs a test that always fails:

```
[ERROR] com.github.pihme.jenkinstestbed.module1.FlakyTest.shouldFailNineInTenInvocations
[ERROR]   Run 1: FlakyTest.shouldFailNineInTenInvocations:15 failed
[ERROR]   Run 2: FlakyTest.shouldFailNineInTenInvocations:15 failed
[ERROR]   Run 3: FlakyTest.shouldFailNineInTenInvocations:15 failed
[ERROR]   Run 4: FlakyTest.shouldFailNineInTenInvocations:15 failed
[ERROR]   Run 5: FlakyTest.shouldFailNineInTenInvocations:15 failed
[ERROR]   Run 6: FlakyTest.shouldFailNineInTenInvocations:15 failed
[ERROR]   Run 7: FlakyTest.shouldFailNineInTenInvocations:15 failed
[ERROR]   Run 8: FlakyTest.shouldFailNineInTenInvocations:15 failed
[ERROR]   Run 9: FlakyTest.shouldFailNineInTenInvocations:15 failed
[ERROR]   Run 10: FlakyTest.shouldFailNineInTenInvocations:15 failed
[ERROR]   Run 11: FlakyTest.shouldFailNineInTenInvocations:15 failed
[ERROR]   Run 12: FlakyTest.shouldFailNineInTenInvocations:15 failed
[ERROR]   Run 13: FlakyTest.shouldFailNineInTenInvocations:15 failed
[ERROR]   Run 14: FlakyTest.shouldFailNineInTenInvocations:15 failed
[ERROR]   Run 15: FlakyTest.shouldFailNineInTenInvocations:15 failed
[ERROR]   Run 16: FlakyTest.shouldFailNineInTenInvocations:15 failed
```

I will continue working on this myself.
But I will also appreciate some help here.

--- a/.ci/scripts/distribution/it-java.sh
+++ b/.ci/scripts/distribution/it-java.sh
@@ -5,9 +5,23 @@ export JAVA_TOOL_OPTIONS="$JAVA_TOOL_OPTIONS -XX:MaxRAMFraction=$((LIMITS_CPU))"
 
 tmpfile=$(mktemp)
 
-mvn -o -B --fail-never -T$LIMITS_CPU -s ${MAVEN_SETTINGS_XML} verify -P skip-unstable-ci,parallel-tests -pl qa/integration-tests -pl upgrade-tests -DtestMavenId=2 | tee ${tmpfile}
+mvn -o -B --fail-never -T$LIMITS_CPU -s ${MAVEN_SETTINGS_XML} verify -P skip-unstable-ci,parallel-tests -pl qa/integration-tests -pl upgrade-tests -DtestMavenId=2 -Dsurefire.rerunFailingTestsCount=5 | tee ${tmpfile}
 
 status=${PIPESTATUS[0]}
+
+if grep -q "\[WARNING\] Flakes:" ${tmpfile}; then
+
+  tmpfile2=$(mktemp)
+
+  awk '/^\[WARNING\] Flakes:.*$/{flag=1}/^\[ERROR\] Tests run:.*Flakes: [0-9]*$/{print;flag=0}flag' ${tmpfile} > ${tmpfile2}
+
+  grep "\[ERROR\]   Run 1: " ${tmpfile} | awk '{print $4}' >> ./target/FlakyTests.txt
+
+  echo ERROR: Flaky Tests detected>&2
+  rm test-it.txt
+  rm flaky-tests-log.txt
+  exit 1
+fi
 
 if [[ $status != 0 ]]; then
   exit $status;

--- a/.ci/scripts/distribution/test-java.sh
+++ b/.ci/scripts/distribution/test-java.sh
@@ -4,9 +4,22 @@ export JAVA_TOOL_OPTIONS="$JAVA_TOOL_OPTIONS -XX:MaxRAMFraction=$((LIMITS_CPU))"
 
 tmpfile=$(mktemp)
 
-mvn -o -B --fail-never -T$LIMITS_CPU -s ${MAVEN_SETTINGS_XML} verify -P skip-unstable-ci,parallel-tests -Dzeebe.it.skip -DtestMavenId=1 | tee ${tmpfile}
+mvn -o -B --fail-never -T$LIMITS_CPU -s ${MAVEN_SETTINGS_XML} verify -P skip-unstable-ci,parallel-tests -Dzeebe.it.skip -DtestMavenId=1 -Dsurefire.rerunFailingTestsCount=5 | tee ${tmpfile}
 
 status=${PIPESTATUS[0]}
+
+if grep -q "\[WARNING\] Flakes:" ${tmpfile}; then
+
+  tmpfile2=$(mktemp)
+
+  awk '/^\[WARNING\] Flakes:.*$/{flag=1}/^\[ERROR\] Tests run:.*Flakes: [0-9]*$/{print;flag=0}flag' ${tmpfile} > ${tmpfile2}
+
+  grep "\[ERROR\]   Run 1: " ${tmpfile2} | awk '{print $4}' >> ./target/FlakyTests.txt
+
+  echo ERROR: Flaky Tests detected>&2
+
+  exit 1
+fi
 
 if [[ $status != 0 ]]; then
   exit $status;

--- a/.ci/scripts/distribution/test-java8.sh
+++ b/.ci/scripts/distribution/test-java8.sh
@@ -1,15 +1,29 @@
 #!/bin/bash -eux
 
-
 export JAVA_TOOL_OPTIONS="$JAVA_TOOL_OPTIONS -XX:MaxRAMFraction=$((LIMITS_CPU))"
 
 mvn -v
 
 tmpfile=$(mktemp)
 
-mvn -o -B --fail-never -T$LIMITS_CPU -s ${MAVEN_SETTINGS_XML} verify -pl clients/java -DtestMavenId=3 | tee ${tmpfile}
+mvn -o -B --fail-never -T$LIMITS_CPU -s ${MAVEN_SETTINGS_XML} verify -pl clients/java -DtestMavenId=3 -Dsurefire.rerunFailingTestsCount=5 | tee ${tmpfile}
 
 status=${PIPESTATUS[0]}
+
+
+if grep -q "\[WARNING\] Flakes:" ${tmpfile}; then
+
+  tmpfile2=$(mktemp)
+
+  awk '/^\[WARNING\] Flakes:.*$/{flag=1}/^\[ERROR\] Tests run:.*Flakes: [0-9]*$/{print;flag=0}flag' ${tmpfile} > ${tmpfile2}
+
+  grep "\[ERROR\]   Run 1: " ${tmpfile2} | awk '{print $4}' >> ./target/FlakyTests.txt
+
+  echo ERROR: Flaky Tests detected>&2
+
+  exit 1
+fi
+
 
 if [[ $status != 0 ]]; then
   exit $status;

--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -189,6 +189,11 @@ pipeline {
             post {
                 failure {
                     archive "**/*/surefire-reports/*-output.txt"
+                    script {
+                      if (fileExists('./target/FlakyTests.txt')) {
+                          currentBuild.description = "Flaky Tests: <br>" + readFile('./target/FlakyTests.txt').split('\n').join('<br>')
+                      }
+                    }
                 }
             }
         }

